# Week 2 Assignment

## Python Introduction
## Variables, Data Types, Conditions & Functions
## Introduction to data structures (lists, tuples, dictionaries)

## Question 1

In [2]:
name = input("Enter your name: ")
age = int(input("Enter your age: "))
email = input("Enter your email: ")
favorite_number = int(input("Enter your favorite number: "))

user_data = {
    "name": name,
    "age": age,
    "email": email,
    "favorite_number": favorite_number
}

if "@" in email and "." in email:
    valid_email = True
else:
    valid_email = False

if valid_email:
    print(f"Hello {name}, you are {age} years old, your email is {email}, and your favorite number is {favorite_number}.")
else:
    print("Invalid email format. Please try again!")

Enter your name: Murtaza
Enter your age: 22
Enter your email: murtazakazmi555@gmail.com
Enter your favorite number: 5
Hello Murtaza, you are 22 years old, your email is murtazakazmi555@gmail.com, and your favorite number is 5.


# User Input and Validation

## Prompt User for Input:
- Ask for name, age, email, and favorite number.
- Convert age and favorite number to integers.

## Store Inputs in a Dictionary:
- Store name, age, email, and favorite number in `user_data` dictionary.

## Validate Email Format:
- Check if email contains both "@" and ".".
- Set `valid_email` to `True` if valid, otherwise `False`.

## Display Message:
- If `valid_email` is `True`, print user's details.
- If `valid_email` is `False`, print an error message.


## Question 2

In [3]:
def Is_even(number):
   
    if number % 2 == 0:
        print(f"{number} is even.")
        return True
    else:
        print(f"{number} is odd.")
        return False
    
    
Is_even(4)
Is_even(3)
Is_even(10)

4 is even.
3 is odd.
10 is even.


True

# Function Definition and Usage

## Define `Is_even` Function:
- Create a function named `Is_even` that takes one parameter, `number`.
- Check if `number` is even by using the modulus operator `%`.

## Check if the Number is Even:
- If `number % 2 == 0`, the number is even:
  - Print that the number is even.
  - Return `True`.
- Else, the number is odd:
  - Print that the number is odd.
  - Return `False`.

## Call `Is_even` Function:
- Call `Is_even(4)`:
  - Output: `4 is even.`
  - Return value: `True`.
- Call `Is_even(3)`:
  - Output: `3 is odd.`
  - Return value: `False`.
- Call `Is_even(10)`:
  - Output: `10 is even.`
  - Return value: `True`.


## Question 3

In [4]:
def convert_temperature(temp, scale):
    
    if scale.upper() == "C": 
        converted_temp = (temp * 9/5) + 32
        print(f"{temp}°C is equal to {converted_temp}°F.")
        return converted_temp
    elif scale.upper() == "F":  
        converted_temp = (temp - 32) * 5/9
        print(f"{temp}°F is equal to {converted_temp}°C.")
        return converted_temp
    else:
        print("Invalid scale. Please enter 'C' for Celsius or 'F' for Fahrenheit.")
        return None
    
    
convert_temperature(25, "C")
convert_temperature(80, "F")
convert_temperature(30, "K")  # Invalid scale

25°C is equal to 77.0°F.
80°F is equal to 26.666666666666668°C.
Invalid scale. Please enter 'C' for Celsius or 'F' for Fahrenheit.


# Function Definition and Usage

## Define `convert_temperature` Function:
- Create a function named `convert_temperature` that takes two parameters: `temp` (temperature) and `scale` (target scale).

## Convert Temperature:
- If `scale.upper() == "C"`:
  - Convert Celsius to Fahrenheit using the formula `(temp * 9/5) + 32`.
  - Print the result and return the converted temperature.
- If `scale.upper() == "F"`:
  - Convert Fahrenheit to Celsius using the formula `(temp - 32) * 5/9`.
  - Print the result and return the converted temperature.
- If the scale is not 'C' or 'F':
  - Print an error message indicating an invalid scale.
  - Return `None`.

## Call `convert_temperature` Function:
- Call `convert_temperature(25, "C")`:
  - Output: `25°C is equal to 77.0°F.`
  - Return value: `77.0`.
- Call `convert_temperature(80, "F")`:
  - Output: `80°F is equal to 26.666666666666668°C.`
  - Return value: `26.666666666666668`.
- Call `convert_temperature(30, "K")`:
  - Output: `Invalid scale. Please enter 'C' for Celsius or 'F' for Fahrenheit.`
  - Return value: `None`.


## Question 4

In [5]:
def find_max_min(numbers_list):
   
    max_num = max(numbers_list)
    min_num = min(numbers_list)
    return max_num, min_num

#  enter 5 numbers
numbers = []
for i in range(5):
    num = float(input(f"Enter number {i+1}: "))
    numbers.append(num)

max_num, min_num = find_max_min(numbers)
print(f"Maximum number: {max_num}")
print(f"Minimum number: {min_num}")

Enter number 1: 1
Enter number 2: 4
Enter number 3: 66
Enter number 4: 22
Enter number 5: 55
Maximum number: 66.0
Minimum number: 1.0


# Function Definition and Usage

## Define `find_max_min` Function:
- Create a function named `find_max_min` that takes one parameter, `numbers_list`.
- Calculate the maximum number using the `max` function.
- Calculate the minimum number using the `min` function.
- Return the maximum and minimum numbers.

## Prompt User to Enter 5 Numbers:
- Initialize an empty list named `numbers`.
- Use a `for` loop to prompt the user to enter 5 numbers:
  - Convert each input to a float.
  - Append each number to the `numbers` list.

## Find and Display Maximum and Minimum Numbers:
- Call `find_max_min(numbers)` to get the maximum and minimum numbers.
- Print the maximum number.
- Print the minimum number.


In [6]:
def find_max_min(numbers_list):
    max_num = numbers_list[0]
    min_num = numbers_list[0]
    for num in numbers_list:
        if num > max_num:
            max_num = num
        elif num < min_num:
            min_num = num
    return max_num, min_num

#  enter 5 numbers
numbers = []
for i in range(5):
    num = float(input(f"Enter number {i+1}: "))
    numbers.append(num)

max_num, min_num = find_max_min(numbers)
print(f"Maximum number: {max_num}")
print(f"Minimum number: {min_num}")

Enter number 1: 55
Enter number 2: 33
Enter number 3: 22
Enter number 4: 1
Enter number 5: 99
Maximum number: 99.0
Minimum number: 1.0


# Function Definition and Usage

## Define `find_max_min` Function:
- Create a function named `find_max_min` that takes one parameter, `numbers_list`.
- Initialize `max_num` and `min_num` with the first element of `numbers_list`.
- Iterate through `numbers_list` to find the maximum and minimum numbers:
  - Update `max_num` if the current number is greater than `max_num`.
  - Update `min_num` if the current number is less than `min_num`.
- Return the maximum and minimum numbers.

## Prompt User to Enter 5 Numbers:
- Initialize an empty list named `numbers`.
- Use a `for` loop to prompt the user to enter 5 numbers:
  - Convert each input to a float.
  - Append each number to the `numbers` list.

## Find and Display Maximum and Minimum Numbers:
- Call `find_max_min(numbers)` to get the maximum and minimum numbers.
- Print the maximum number.
- Print the minimum number.


## Question 5

In [7]:
students = []
for i in range(3):
    name = input(f"Enter name of student {i+1}: ")
    age = int(input(f"Enter age of student {i+1}: "))
    grade = input(f"Enter grade of student {i+1}: ")
    students.append((name, age, grade))

students_dict = {name: (age, grade) for name, age, grade in students}

# Display output
print("Student Details:")
for name, (age, grade) in students_dict.items():
    print(f"Name: {name}, Age: {age}, Grade: {grade}")

Enter name of student 1: murtaza
Enter age of student 1: 22
Enter grade of student 1: A
Enter name of student 2: Ali
Enter age of student 2: 21
Enter grade of student 2: B
Enter name of student 3: Haseeb
Enter age of student 3: 23
Enter grade of student 3: A+
Student Details:
Name: murtaza, Age: 22, Grade: A
Name: Ali, Age: 21, Grade: B
Name: Haseeb, Age: 23, Grade: A+


# Prompt User to Enter Details of 3 Students

## Initialize `students` List:
- Create an empty list named `students` to store details of students.

## Loop to Collect Student Details:
- Use a `for` loop to prompt the user to enter details for 3 students:
  - Ask for the name of the student.
  - Ask for the age of the student (input is converted to an integer).
  - Ask for the grade of the student.
  - Append a tuple `(name, age, grade)` to the `students` list.

## Convert List of Tuples to Dictionary:
- Create a dictionary named `students_dict` using a dictionary comprehension:
  - Use the name as the key.
  - Use a tuple `(age, grade)` as the value.

## Display Student Details:
- Print "Student Details:" to indicate the start of output.
- Iterate through `students_dict` using `.items()` to print each student's details:
  - Print the student's name, age, and grade formatted accordingly.


## Question 6

In [12]:
def update_inventory(inventory_dict, item, quantity):
   
    if item in inventory_dict:
        new_quantity = inventory_dict[item] + quantity
        if new_quantity >= 0:
            inventory_dict[item] = new_quantity
        else:
            print(f"Cannot remove more {item} than available.")
    else:
        print(f"{item} not found in inventory.")
    return inventory_dict

inventory = {
    "Apple": 10,
    "Banana": 20,
    "Cherry": 15,
    "Date": 12,
    "Elderberry": 18
}

for i in range(3):
    item = input(f"Enter item {i+1} to update: ")
    quantity = int(input(f"Enter quantity to add/remove ({item}): "))
    inventory = update_inventory(inventory, item, quantity)

print("Updated Inventory:")
for item, quantity in inventory.items():
    print(f"{item}: {quantity}")

Enter item 1 to update: Apple
Enter quantity to add/remove (Apple): 5
Enter item 2 to update: Banana
Enter quantity to add/remove (Banana): 5
Enter item 3 to update: Date
Enter quantity to add/remove (Date): 7
Updated Inventory:
Apple: 15
Banana: 25
Cherry: 15
Date: 19
Elderberry: 18


# Function Definition and Usage

## Define `update_inventory` Function:
- Create a function named `update_inventory` that takes three parameters:
  - `inventory_dict`: Dictionary representing current inventory.
  - `item`: Name of the item to update.
  - `quantity`: Quantity to add or remove for the item.
- Check if `item` exists in `inventory_dict`:
  - If yes, calculate `new_quantity` by adding `quantity` to the current quantity.
  - If `new_quantity` is non-negative, update `inventory_dict` with `new_quantity`.
  - If `new_quantity` is negative, print an error message indicating insufficient quantity to remove.
- If `item` does not exist in `inventory_dict`, print an error message indicating item not found.
- Return the updated `inventory_dict`.

## Initialize Inventory Dictionary:
- Create an inventory dictionary `inventory` with initial items and quantities.

## Prompt User to Update Inventory:
- Use a `for` loop to prompt the user to update inventory for 3 items:
  - Ask for the item name.
  - Ask for the quantity to add or remove for the item (input converted to an integer).
  - Call `update_inventory` function to update `inventory` based on user input.

## Display Updated Inventory:
- Print "Updated Inventory:" to indicate the start of output.
- Iterate through `inventory` dictionary using `.items()` to print each item and its updated quantity.


